Mathlib/RingTheory/MvPolynomial/NewtonIdentities.lean

```diff
@@ -0,0 +1,265 @@
+  symmetric polynomial in terms of lower degree elementary symmetric polynomials and power sums.
+
+## References
+
+See [zeilberger1984] for the combinatorial proof of Newton's identities.
+-/
+
+open Equiv (Perm)
+
+open BigOperators MvPolynomial
+
+noncomputable section
+
+namespace MvPolynomial
+
+namespace NewtonIdentities
+
+open Finset Nat
+
+variable (σ : Type _) [Fintype σ] [DecidableEq σ] (R : Type _) [CommRing R] [NoZeroDivisors R]
+  [CharZero R]
+
+private def pairs (k : ℕ) : Finset (Finset σ × σ) :=
+  univ.filter (fun t ↦ card t.fst ≤ k ∧ (card t.fst = k → t.snd ∈ t.fst))
+
+@[simp]
+private lemma mem_pairs (k : ℕ) (t : Finset σ × σ) :
+    t ∈ pairs σ k ↔ card t.fst ≤ k ∧ (card t.fst = k → t.snd ∈ t.fst) := by
+  simp [pairs]
+
+private def weight (k : ℕ) (t : Finset σ × σ) : MvPolynomial σ R :=
+  (-1) ^ card t.fst * ((∏ a in t.fst, X a) * X t.snd ^ (k - card t.fst))
+
+private def pairMap (t : Finset σ × σ) : Finset σ × σ :=
+  if h : t.snd ∈ t.fst then (t.fst.erase t.snd, t.snd) else (t.fst.cons t.snd h, t.snd)
+
+private lemma pairMap_of_snd_mem_fst {t : Finset σ × σ} (h : t.snd ∈ t.fst) :
+    pairMap σ t = (t.fst.erase t.snd, t.snd) := by
+  simp [pairMap, h]
+
+private lemma pairMap_of_snd_nmem_fst {t : Finset σ × σ} (h : t.snd ∉ t.fst) :
+    pairMap σ t = (t.fst.cons t.snd h, t.snd) := by
+  simp [pairMap, h]
+
+private theorem pairMap_mem_pairs (t : Finset σ × σ) (h : t ∈ pairs σ k) :
+    pairMap σ t ∈ pairs σ k := by
+  rw [mem_pairs] at h ⊢
+  rcases (em (t.snd ∈ t.fst)) with h1 | h1
+  · rw [pairMap_of_snd_mem_fst σ h1]
+    simp only [h1, implies_true, and_true] at h
+    simp only [card_erase_of_mem h1, tsub_le_iff_right, mem_erase, ne_eq, h1]
+    refine ⟨le_step h, ?_⟩
+    by_contra h2
+    rw [← h2] at h
+    exact not_le_of_lt (sub_lt (card_pos.mpr ⟨t.snd, h1⟩) zero_lt_one) h
+  · rw [pairMap_of_snd_nmem_fst σ h1]
+    simp only [h1] at h
+    simp only [card_cons, mem_cons, true_or, implies_true, and_true]
+    exact (le_iff_eq_or_lt.mp h.left).resolve_left h.right
+
+@[simp]
+private theorem pairMap_involutive : (pairMap σ).Involutive := by
+  intro t
+  rw [pairMap, pairMap]
+  split_ifs with h1 h2 h3
+  · simp at h2
+  · simp [insert_erase h1]
+  · simp_all
+  · simp at h3
+
+private theorem weight_add_weight_pairMap (t : Finset σ × σ) (h : t ∈ pairs σ k) :
+    weight σ R k t + weight σ R k (pairMap σ t) = 0 := by
+  rw [weight, weight]
+  rw [mem_pairs] at h
+  have h2 (n : ℕ) : -(-1 : MvPolynomial σ R) ^ n = (-1) ^ (n + 1) := by
+    rw [← neg_one_mul ((-1 : MvPolynomial σ R) ^ n), pow_add, pow_one, mul_comm]
+  rcases (em (t.snd ∈ t.fst)) with h1 | h1
+  · rw [pairMap_of_snd_mem_fst σ h1]
+    simp only [← prod_erase_mul t.fst (fun j ↦ (X j : MvPolynomial σ R)) h1,
+      mul_assoc (∏ a in erase t.fst t.snd, X a), card_erase_of_mem h1]
+    nth_rewrite 1 [← pow_one (X t.snd)]
+    simp only [← pow_add, add_comm]
+    have h3 : card t.fst ≥ 1 := lt_iff_add_one_le.mp (card_pos.mpr ⟨t.snd, h1⟩)
```

Comment:
Apart from rare exceptions, we always use `<` and `≤` instead of `>` and `≥` since it's occasionally awkward having syntactically different but definitionally equal things floating around:
```suggestion
    have h3 : 1 ≤ card t.fst := lt_iff_add_one_le.mp (card_pos.mpr ⟨t.snd, h1⟩)
```